### PR TITLE
Add regenerate bundle request type

### PR DESCRIPTION
**Pull request by lcarva**
Introduce the new API endpoint /api/v1/builds/regenerate-bundle.
This is used to rebuild an operator bundle image.
In the future, modifications will be applied to the bundle image as part
of the build process.

Ticket-ID: CLOUDBLD-520

Signed-off-by: Luiz Carvalho <lucarval@redhat.com>

### CODE REVIEW

File: iib/web/models.py [hunk 1]
```diff
@@ -658,6 +662,80 @@ def to_json(self, verbose=True):
...
+        request_kwargs = deepcopy(kwargs)
+
+        validate_request_params(
+            request_kwargs, required_params={'from_bundle_image'}, optional_params={'organization'},
+        )
+
+        # Validate organization is correctly provided
+        organization = request_kwargs.get('organization')
+        if organization and not isinstance(organization, str):
+            raise ValidationError(f'"organization" must be a string')
```
Comment (yashvardhannanavati) [inline iib/web/models.py:701]: f' "organization" must be a string' => ' "organization" must be a string' ?

File: iib/web/models.py [hunk 2]
```diff
@@ -658,6 +662,80 @@ def to_json(self, verbose=True):
...
+        # Validate from_bundle_image is correctly provided
+        from_bundle_image = request_kwargs.pop('from_bundle_image')
+        if not isinstance(from_bundle_image, str):
+            raise ValidationError('"from_bundle_image" must be a string')
+
+        request_kwargs['from_bundle_image'] = Image.get_or_create(
+            pull_specification=from_bundle_image
+        )
+
+        # current_user.is_authenticated is only ever False when auth is disabled
```
Comment (yashvardhannanavati) [inline iib/web/models.py:712]: Does it make sense to move this before you add the from_bundle_image param to the db? That would avoid unnecessary database operations. Thoughts?

File: iib/web/models.py [hunk 3]
```diff
@@ -658,6 +662,80 @@ def to_json(self, verbose=True):
...
+        if current_user.is_authenticated:
+            request_kwargs['user'] = current_user
+
+        request = cls(**request_kwargs)
+        request.add_state('in_progress', 'The request was initiated')
+        return request
+
+    def to_json(self, verbose=True):
+        """
+        Provide the JSON representation of a regenerate-bundle build request.
```
Comment (yashvardhannanavati) [inline iib/web/models.py:722]: Provide the JSON representation of a regenerate-bundle build request. => Provide the JSON representation of a "regenerate-bundle" build request.

File: iib/workers/tasks/build.py [hunk 4]
```diff
@@ -227,12 +234,12 @@ def _get_image_arches(pull_spec):
     return arches
 
 
-def get_rebuilt_index_image(request_id):
+def get_rebuilt_image_pull_spec(request_id):
     """
-    Generate the pull specification of the index image rebuilt by IIB.
+    Generate the pull specification of the image rebuilt by IIB.
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/build.py:239]: How about keeping it consistent throughout and having `container image` everywhere?

File: iib/workers/tasks/build.py [hunk 5]
```diff
@@ -227,12 +234,12 @@ def _get_image_arches(pull_spec):
...
 
     :param int request_id: the ID of the IIB build request
-    :return: pull specification of the rebuilt index image
+    :return: pull specification of the rebuilt image
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/build.py:242]: Same as the previous comment of keeping `container image` consistent.

File: iib/workers/tasks/build.py [hunk 6]
```diff
@@ -634,3 +641,63 @@ def handle_rm_request(
...
+
+    arches_str = ', '.join(sorted(arches))
+    log.debug('Set to regenerate the bundle image for the following arches: %s', arches_str)
+
+    payload = {
+        'from_bundle_image_resolved': from_bundle_image_resolved,
+        'state': 'in_progress',
+        'state_reason': f'Regenerating the bundle image for the following arches: {arches_str}',
+    }
+    exc_msg = 'Failed setting the resolved from_bundle_image on the request'
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/build.py:674]: `'Failed setting the resolved from_bundle_image on the request'` => `'Failed setting the resolved "from_bundle_image" on the request'`

**Comment by yashvardhannanavati** [comment]
@lcarva just out of curiosity, can I please ask why would it just regenerate a single bundle image at once and not multiple?

**Comment by lcarva** [comment]
Oops I forgot to update the API spec. I'll do that.

File: iib/web/models.py (hunk 2 above)
Comment (lcarva) [reply iib/web/models.py:712]: I don't understand the question. If the instance of IIB does not have authentication enabled, then `user` is not set. That's a valid scenario, e.g. local dev environment with docker-compose. In either case, the same amount of database operations are performed.

File: iib/web/models.py (hunk 1 above)
Comment (lcarva) [reply iib/web/models.py]: Thanks for catching! Bad copy and past :stuck_out_tongue:

File: iib/workers/tasks/build.py (hunk 4 above)
Comment (lcarva) [reply iib/workers/tasks/build.py]: I added as a separate commit because I had to touch other methods in this file.

**Comment by lcarva** [comment]
> @lcarva just out of curiosity, can I please ask why would it just regenerate a single bundle image at once and not multiple?

It seemed more natural to produce a single image per request.

**Comment by lcarva** [comment]
> Oops I forgot to update the API spec. I'll do that.

Added.

File: iib/web/static/api_v1.yaml [hunk 7]
```diff
@@ -166,6 +172,7 @@ paths:
         content:
           application/json:
             schema:
+              # TODO: update to include Regenerate Bundle attrs
```
[the remark this answers is not in the document]
Comment (lcarva) [reply iib/web/static/api_v1.yaml:175]: I did. I can remove it if it adds noise and no benefit.

[the remark this answers is not in the document]
Comment (mprahl) [reply iib/web/static/api_v1.yaml:175]: I might not understand the comment then. Why isn't it included in this PR?

[the remark this answers is not in the document]
Comment (lcarva) [reply iib/web/static/api_v1.yaml:175]: It'll be included in a separate PR done in a different jira story.

File: tests/test_web/test_migrations.py [hunk 8]
```diff
@@ -35,8 +39,46 @@ def test_migrate_to_polymorphic_requests(app, auth_env, client, db):
...
+
+
+def test_abort_when_downgrading_from_regenerate_bundle_request(app, auth_env, client, db):
+    """Verify downgrade is prevented if "regenerate-bundle" requests exist."""
+    total_requests = 20
+    # flask_login.current_user is used in Request*.from_json which requires a request context
+    with app.test_request_context(environ_base=auth_env):
+        # Generate some data to verify migration
+        for i in range(total_requests):
+            request_class = random.choice((RequestAdd, RequestRm, RequestRegenerateBundle))
```
[the remark this answers is not in the document]
Comment (lcarva) [reply tests/test_web/test_migrations.py:57]: It is. `(2/3) ^ 20` chance? I added an assertion to verify the assumption that at least one of these objects exist. Given the unlikelihood of this happening, I added an assertion message that should guide our future selves.

**Comment by lcarva** [comment]
@mprahl, I resolved the comments where I applied changes, but left some comments open because I'm not clear if there's any change required.

File: tests/test_web/test_migrations.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mprahl) [reply tests/test_web/test_migrations.py:57]: @lcarva why not just always add a `RequestRegenerateBundle` and then continue with adding 19 random requests?

**Comment by mprahl** [comment]
@lcarva I replied to one of your comments. Once that is addressed, this looks good to me.

File: tests/test_web/test_migrations.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (lcarva) [reply tests/test_web/test_migrations.py]: Good suggestion! Done.

File: iib/web/models.py [hunk 9]
```diff
@@ -350,6 +362,15 @@ def _state_to_json(state):
 
         return rv
 
+    def get_modifiable_keys(self):
```
Comment (yashvardhannanavati) [inline iib/web/models.py:365]: Optional: how about renaming `get_modifiable_keys` to `get_mutable_keys`?

File: iib/web/models.py (hunk 2 above)
Comment (yashvardhannanavati) [reply iib/web/models.py:769]: my bad, I misinterpreted something. If login is enabled and the user is not authenticated, it wouldn't reach here in the first place.

File: iib/web/models.py (hunk 9 above)
Comment (lcarva) [reply iib/web/models.py]: Great suggestion! Done.

**Pull request merged**